File: src/protocols/matrix/native/map.ts
import * as sdk from '@unomed/react-native-matrix-sdk';

import type { MxContent, MxEvent, MxMedia, MxMembership, MxMembershipChange, MxRole } from '../api';

export function mapMembership(membership: sdk.Membership): MxMembership {
  switch (membership) {
    case sdk.Membership.Invited:
      return 'invited';
    case sdk.Membership.Joined:
      return 'joined';
    case sdk.Membership.Banned:
      return 'banned';
    case sdk.Membership.Knocked:
      return 'knocked';
    default:
      return 'left';
  }
}

export function mapRole(role: sdk.RoomMemberRole): MxRole {
  switch (role) {
    case sdk.RoomMemberRole.Creator:
      return 'owner';
    case sdk.RoomMemberRole.Administrator:
    case sdk.RoomMemberRole.Moderator:
      return 'admin';
    default:
      return 'member';
  }
}

export function nameOf(profile: sdk.ProfileDetails): string | undefined {
  return profile.tag === sdk.ProfileDetails_Tags.Ready
    ? (profile.inner.displayName ?? undefined)
    : undefined;
}

export function sendState(state: sdk.EventSendState | undefined): MxEvent['status'] {
  if (!state) return 'sent';
  switch (state.tag) {
    case sdk.EventSendState_Tags.Sent:
      return 'sent';
    case sdk.EventSendState_Tags.SendingFailed:
      return 'failed';
    default:
      return 'sending';
  }
}

export function mapContent(content: sdk.TimelineItemContent): MxContent | null {
  switch (content.tag) {
    case sdk.TimelineItemContent_Tags.MsgLike:
      return mapMsgLike(content.inner.content.kind);
    case sdk.TimelineItemContent_Tags.RoomMembership: {
      const change = mapMembershipChange(content.inner.change);
      return change
        ? {
            kind: 'membership',
            change,
            user: content.inner.userId,
            userName: content.inner.userDisplayName,
          }
        : null;
    }
    case sdk.TimelineItemContent_Tags.State:
      return mapState(content.inner.content);
    default:
      return null;
  }
}

function mapMsgLike(kind: sdk.MsgLikeKind): MxContent | null {
  switch (kind.tag) {
    case sdk.MsgLikeKind_Tags.Message:
      return mapMessage(kind.inner.content.msgType);
    case sdk.MsgLikeKind_Tags.Sticker:
      return { kind: 'sticker', body: kind.inner.body };
    case sdk.MsgLikeKind_Tags.Poll:
      return {
        kind: 'poll',
        question: kind.inner.question,
        answers: kind.inner.answers,
        votes: Object.fromEntries(kind.inner.votes),
        maxSelections: Number(kind.inner.maxSelections),
        closed: kind.inner.endTime !== undefined,
      };
    case sdk.MsgLikeKind_Tags.Redacted:
      return { kind: 'redacted' };
    case sdk.MsgLikeKind_Tags.UnableToDecrypt:
      return { kind: 'undecryptable' };
    case sdk.MsgLikeKind_Tags.LiveLocation:
      return { kind: 'location' };
    default:
      return null;
  }
}

function textOf(content: { body: string; formatted?: sdk.FormattedBody }) {
  const html =
    content.formatted?.format.tag === sdk.MessageFormat_Tags.Html
      ? content.formatted.body
      : undefined;
  return { body: content.body, html };
}

function mapMessage(type: sdk.MessageType): MxContent | null {
  switch (type.tag) {
    case sdk.MessageType_Tags.Text:
      return { kind: 'text', ...textOf(type.inner.content) };
    case sdk.MessageType_Tags.Notice:
      return { kind: 'text', ...textOf(type.inner.content), msgtype: 'notice' };
    case sdk.MessageType_Tags.Emote:
      return { kind: 'text', ...textOf(type.inner.content), msgtype: 'emote' };
    case sdk.MessageType_Tags.Image: {
      const c = type.inner.content;
      return {
        kind: 'image',
        ...media(c.source, c.filename, c.info?.mimetype, c.info?.size),
        width: c.info?.width === undefined ? undefined : Number(c.info.width),
        height: c.info?.height === undefined ? undefined : Number(c.info.height),
        caption: c.caption,
      };
    }
    case sdk.MessageType_Tags.File: {
      const c = type.inner.content;
      return {
        kind: 'file',
        ...media(c.source, c.filename, c.info?.mimetype, c.info?.size),
        caption: c.caption,
      };
    }
    case sdk.MessageType_Tags.Audio: {
      const c = type.inner.content;
      const durationMs = c.info?.duration ?? c.audio?.duration;
      return {
        kind: 'audio',
        ...media(c.source, c.filename, c.info?.mimetype, c.info?.size),
        durationMs: durationMs === undefined ? undefined : Math.round(durationMs),
        voice: c.voice !== undefined,
      };
    }
    case sdk.MessageType_Tags.Video: {
      const c = type.inner.content;
      return {
        kind: 'video',
        ...media(c.source, c.filename, c.info?.mimetype, c.info?.size),
        width: c.info?.width === undefined ? undefined : Number(c.info.width),
        height: c.info?.height === undefined ? undefined : Number(c.info.height),
        durationMs: c.info?.duration,
        caption: c.caption,
      };
    }
    case sdk.MessageType_Tags.Location:
      return { kind: 'location' };
    default:
      return null;
  }
}

function media(
  source: sdk.MediaSourceLike,
  name: string,
  mimeType: string | undefined,
  size: bigint | undefined
): MxMedia {
  return {
    source: source.toJson(),
    name,
    mimeType,
    size: size === undefined ? undefined : Number(size),
  };
}

function mapMembershipChange(change: sdk.MembershipChange | undefined): MxMembershipChange | null {
  switch (change) {
    case sdk.MembershipChange.Joined:
    case sdk.MembershipChange.InvitationAccepted:
      return 'joined';
    case sdk.MembershipChange.Left:
      return 'left';
    case sdk.MembershipChange.Invited:
      return 'invited';
    case sdk.MembershipChange.Kicked:
    case sdk.MembershipChange.KickedAndBanned:
      return 'kicked';
    case sdk.MembershipChange.Banned:
      return 'banned';
    case sdk.MembershipChange.Unbanned:
      return 'unbanned';
    case sdk.MembershipChange.InvitationRejected:
      return 'invitationRejected';
    case sdk.MembershipChange.InvitationRevoked:
      return 'invitationRevoked';
    default:
      return null;
  }
}

function mapState(state: sdk.OtherState): MxContent | null {
  switch (state.tag) {
    case sdk.OtherState_Tags.RoomName:
      return { kind: 'state', change: 'name', value: state.inner.name ?? undefined };
    case sdk.OtherState_Tags.RoomTopic:
      return { kind: 'state', change: 'topic', value: state.inner.topic ?? undefined };
    case sdk.OtherState_Tags.RoomAvatar:
      return { kind: 'state', change: 'avatar' };
    case sdk.OtherState_Tags.RoomCreate:
      return { kind: 'state', change: 'created' };
    case sdk.OtherState_Tags.RoomEncryption:
      return { kind: 'state', change: 'encryption' };
    default:
      return null;
  }
}

export function extensionOf(name: string): string {
  const match = name.match(/\.[A-Za-z0-9]{1,5}$/);
  return match ? match[0].toLowerCase() : '';
}

File: src-tauri/src/matrix/map.rs
use super::*;

pub(super) fn preview_of(latest: LatestEventValue) -> Option<MxPreview> {
    let (timestamp, sender, is_own, profile, content) = match latest {
        LatestEventValue::Remote {
            timestamp,
            sender,
            is_own,
            profile,
            content,
        } => (timestamp, sender, is_own, profile, content),
        LatestEventValue::Local {
            timestamp,
            sender,
            profile,
            content,
            ..
        } => (timestamp, sender, true, profile, content),
        _ => return None,
    };
    Some(MxPreview {
        sender: sender.to_string(),
        sender_name: name_of(&profile),
        timestamp: timestamp.0.into(),
        is_own,
        content: map_content(&content)?,
    })
}

/// Local echoes and events the app never shows are skipped; our own sends surface once the homeserver has them.
pub(super) fn to_mx_event(room_id: &RoomId, item: &TimelineItem) -> Option<MxEvent> {
    let event = item.as_event()?;
    let event_id = event.event_id()?;
    let content = map_content(event.content())?;
    let msg_like = match event.content() {
        TimelineItemContent::MsgLike(msg) => Some(msg),
        _ => None,
    };
    let reactions: Vec<MxReaction> = msg_like
        .map(|m| {
            m.reactions
                .iter()
                .map(|(key, senders)| MxReaction {
                    key: key.clone(),
                    senders: senders.keys().map(|user| user.to_string()).collect(),
                })
                .collect()
        })
        .unwrap_or_default();
    Some(MxEvent {
        id: event_id.to_string(),
        room_id: room_id.to_string(),
        preview: MxPreview {
            sender: event.sender().to_string(),
            sender_name: name_of(event.sender_profile()),
            timestamp: event.timestamp().0.into(),
            is_own: event.is_own(),
            content,
        },
        status: match event.send_state() {
            None | Some(EventSendState::Sent { .. }) => "sent",
            Some(EventSendState::SendingFailed { .. }) => "failed",
            Some(EventSendState::NotSentYet { .. }) => "sending",
        },
        reply_to: msg_like.and_then(|m| m.in_reply_to.as_ref().map(|r| r.event_id.to_string())),
        reactions: (!reactions.is_empty()).then_some(reactions),
        edited: msg_like.is_some_and(
            |m| matches!(&m.kind, MsgLikeKind::Message(message) if message.is_edited()),
        ),
    })
}

pub(super) fn name_of(
    profile: &TimelineDetails<matrix_sdk_ui::timeline::Profile>,
) -> Option<String> {
    match profile {
        TimelineDetails::Ready(profile) => profile.display_name.clone(),
        _ => None,
    }
}

pub(super) fn media_of(
    source: &matrix_sdk::ruma::events::room::MediaSource,
    name: &str,
    mime_type: Option<&str>,
    size: Option<matrix_sdk::ruma::UInt>,
) -> MxMediaOut {
    MxMediaOut {
        source: serde_json::to_string(source).unwrap_or_default(),
        name: name.to_string(),
        mime_type: mime_type.map(str::to_string),
        size: size.map(Into::into),
    }
}

pub(super) fn map_content(content: &TimelineItemContent) -> Option<MxContent> {
    match content {
        TimelineItemContent::MsgLike(msg) => match &msg.kind {
            MsgLikeKind::Message(message) => map_message(message.msgtype()),
            MsgLikeKind::Sticker(sticker) => Some(MxContent::Sticker {
                body: sticker.content().body.clone(),
            }),
            MsgLikeKind::Poll(poll) => {
                let result = poll.results();
                Some(MxContent::Poll {
                    question: result.question,
                    answers: result
                        .answers
                        .into_iter()
                        .map(|answer| MxPollAnswer {
                            id: answer.id,
                            text: answer.text,
                        })
                        .collect(),
                    votes: result.votes,
                    max_selections: result.max_selections,
                    closed: result.end_time.is_some(),
                })
            }
            MsgLikeKind::Redacted => Some(MxContent::Redacted),
            MsgLikeKind::UnableToDecrypt(_) => Some(MxContent::Undecryptable),
            MsgLikeKind::LiveLocation(_) => Some(MxContent::Location),
            MsgLikeKind::Other(_) => None,
        },
        TimelineItemContent::MembershipChange(change) => {
            let change_str = match change.change()? {
                MembershipChange::Joined | MembershipChange::InvitationAccepted => "joined",
                MembershipChange::Left => "left",
                MembershipChange::Invited => "invited",
                MembershipChange::Kicked | MembershipChange::KickedAndBanned => "kicked",
                MembershipChange::Banned => "banned",
                MembershipChange::Unbanned => "unbanned",
                MembershipChange::InvitationRejected => "invitationRejected",
                MembershipChange::InvitationRevoked => "invitationRevoked",
                _ => return None,
            };
            Some(MxContent::Membership {
                change: change_str,
                user: change.user_id().to_string(),
                user_name: change.display_name(),
            })
        }
        TimelineItemContent::OtherState(state) => {
            let (change, value) = match state.content() {
                AnyOtherStateEventContentChange::RoomName(c) => (
                    "name",
                    match c {
                        StateEventContentChange::Original { content, .. } => {
                            Some(content.name.clone()).filter(|n| !n.is_empty())
                        }
                        StateEventContentChange::Redacted(_) => None,
                    },
                ),
                AnyOtherStateEventContentChange::RoomTopic(c) => (
                    "topic",
                    match c {
                        StateEventContentChange::Original { content, .. } => {
                            Some(content.topic.clone())
                        }
                        StateEventContentChange::Redacted(_) => None,
                    },
                ),
                AnyOtherStateEventContentChange::RoomAvatar(_) => ("avatar", None),
                AnyOtherStateEventContentChange::RoomCreate(_) => ("created", None),
                AnyOtherStateEventContentChange::RoomEncryption(_) => ("encryption", None),
                _ => return None,
            };
            Some(MxContent::State { change, value })
        }
        _ => None,
    }
}

pub(super) fn html_of(formatted: Option<&FormattedBody>) -> Option<String> {
    formatted
        .filter(|f| f.format == MessageFormat::Html)
        .map(|f| f.body.clone())
}

pub(super) fn map_message(msgtype: &MessageType) -> Option<MxContent> {
    Some(match msgtype {
        MessageType::Text(t) => MxContent::Text {
            body: t.body.clone(),
            html: html_of(t.formatted.as_ref()),
            msgtype: None,
        },
        MessageType::Notice(n) => MxContent::Text {
            body: n.body.clone(),
            html: html_of(n.formatted.as_ref()),
            msgtype: Some("notice"),
        },
        MessageType::Emote(e) => MxContent::Text {
            body: e.body.clone(),
            html: html_of(e.formatted.as_ref()),
            msgtype: Some("emote"),
        },
        MessageType::Image(i) => MxContent::Image {
            media: media_of(
                &i.source,
                i.filename(),
                i.info.as_ref().and_then(|info| info.mimetype.as_deref()),
                i.info.as_ref().and_then(|info| info.size),
            ),
            width: i.info.as_ref().and_then(|info| info.width).map(Into::into),
            height: i.info.as_ref().and_then(|info| info.height).map(Into::into),
            caption: i.caption().map(str::to_string),
        },
        MessageType::File(f) => MxContent::File {
            media: media_of(
                &f.source,
                f.filename(),
                f.info.as_ref().and_then(|info| info.mimetype.as_deref()),
                f.info.as_ref().and_then(|info| info.size),
            ),
            caption: f.caption().map(str::to_string),
        },
        MessageType::Audio(a) => MxContent::Audio {
            media: media_of(
                &a.source,
                a.filename(),
                a.info.as_ref().and_then(|info| info.mimetype.as_deref()),
                a.info.as_ref().and_then(|info| info.size),
            ),
            duration_ms: a
                .info
                .as_ref()
                .and_then(|info| info.duration)
                .or_else(|| a.audio.as_ref().map(|audio| audio.duration))
                .map(|d| d.as_millis() as u64),
            voice: a.voice.is_some(),
        },
        MessageType::Video(v) => MxContent::Video {
            media: media_of(
                &v.source,
                v.filename(),
                v.info.as_ref().and_then(|info| info.mimetype.as_deref()),
                v.info.as_ref().and_then(|info| info.size),
            ),
            width: v.info.as_ref().and_then(|info| info.width).map(Into::into),
            height: v.info.as_ref().and_then(|info| info.height).map(Into::into),
            duration_ms: v
                .info
                .as_ref()
                .and_then(|info| info.duration)
                .map(|duration| duration.as_millis() as u64),
            caption: v.caption().map(str::to_string),
        },
        MessageType::Location(_) => MxContent::Location,
        _ => return None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use matrix_sdk::ruma::events::room::message::{
        ImageMessageEventContent, NoticeMessageEventContent, TextMessageEventContent,
    };
    use serde_json::json;

    fn mapped(msgtype: MessageType) -> serde_json::Value {
        serde_json::to_value(map_message(&msgtype).expect("shown")).unwrap()
    }

    #[test]
    fn text_keeps_its_html_and_kind() {
        assert_eq!(
            mapped(MessageType::Text(TextMessageEventContent::html(
                "hi",
                "<b>hi</b>"
            ))),
            json!({ "kind": "text", "body": "hi", "html": "<b>hi</b>" })
        );
        assert_eq!(
            mapped(MessageType::Notice(NoticeMessageEventContent::plain(
                "bot says"
            ))),
            json!({ "kind": "text", "body": "bot says", "msgtype": "notice" })
        );
    }

    #[test]
    fn media_carries_the_source_and_name() {
        let image =
            ImageMessageEventContent::plain("cat.png".into(), "mxc://example.org/abc".into());
        let value = mapped(MessageType::Image(image));
        assert_eq!(value["kind"], "image");
        assert_eq!(value["name"], "cat.png");
        assert_eq!(value["source"], r#"{"url":"mxc://example.org/abc"}"#);
    }
}
